godot-macros/src/docs.rs

```diff
@@ -0,0 +1,224 @@
+        },
+    }
+}
+
+/// `///` is expanded to `#[doc = "…"]`.
+/// This function goes through and extracts the …
+fn siphon_docs_from_attributes(doc: &[Attribute]) -> impl Iterator<Item = String> + '_ {
+    doc.iter()
+        // find #[doc]
+        .filter(|x| x.get_single_path_segment().is_some_and(|x| x == "doc"))
+        // #[doc = "…"]
+        .filter_map(|x| match &x.value {
+            AttributeValue::Equals(_, doc) => Some(doc),
+            _ => None,
+        })
+        .flat_map(|doc| {
+            doc.into_iter().map(|x| {
+                x.to_string()
+                    .trim_start_matches('r')
+                    .trim_start_matches('#')
+                    .trim_start_matches('"')
+                    .trim_end_matches('#')
+                    .trim_end_matches('"')
+                    .to_string()
+            })
+        })
+}
+
+/// Calls [`siphon_docs_from_attributes`] and converts the result to BBCode
+/// for Godot's consumption.
+fn make_docs_from_attributes(doc: &[Attribute]) -> Option<String> {
+    let doc = siphon_docs_from_attributes(doc)
+        .collect::<Vec<_>>()
+        .join("\n");
+    (!doc.is_empty()).then(|| markdown_converter::to_bbcode(&doc))
+}
+
+fn make_signal_docs(signal: &SignalDefinition) -> Option<String> {
+    let name = &signal.signature.name;
+    let params = params(signal.signature.params.iter().filter_map(|(x, _)| match x {
+        FnParam::Receiver(_) => None,
+        FnParam::Typed(y) => Some((&y.name, &y.ty)),
+    }));
+    let desc = make_docs_from_attributes(&signal.external_attributes)?;
+    Some(format!(
+        r#"
+<signal name="{name}">
+  {params}
+  <description>
+  {desc}
+  </description>
+</signal>
+"#
+    ))
+}
+
+fn constant(constant: &Constant) -> Option<String> {
+    let docs = make_docs_from_attributes(&constant.attributes)?;
+    let name = constant.name.to_string();
+    let value = constant
+        .initializer
+        .as_ref()
+        .map(|x| x.to_token_stream().to_string())
+        .unwrap_or("null".into());
+    Some(format!(
+        r#"<constant name="{name}" value="{value}">{docs}</constant>"#
+    ))
+}
+
+pub fn member(member: &Field) -> Option<String> {
+    let docs = make_docs_from_attributes(&member.attributes)?;
+    let name = &member.name;
+    let ty = member.ty.to_token_stream().to_string();
+    let default = member.default.to_token_stream().to_string();
+    Some(format!(
+        r#"<member name="{name}" type="{ty}" default="{default}">{docs}</member>"#
+    ))
+}
+
+fn params<'a, 'b>(params: impl Iterator<Item = (&'a Ident, &'b TypeExpr)>) -> String {
+    params
+        .enumerate()
+        .fold(String::default(), |mut acc, (index, (name, ty))| {
+            use std::fmt::Write;
+            _ = write!(
+                acc,
+                r#"<param index="{index}" name="{name}" type="{ty}" />"#,
+                ty = ty.to_token_stream()
+            );
+            acc
+        })
+}
```

Comment:
Also here, I think a `let mut result = String::new();` followed by a `for` loop is easier to understand.

Reply:
What does clippy not like or suggest instead?

This review would be so much faster if you could provide the relevant information without us always having to ask for it :pensive:

Reply:
If you need `#[allow]`, then rather use the variant with `for` loop. It might also need fewer allocations if a single string is extended.

Reply:
clippy suggests this.

Reply:
we cant use `.extend` or anything because we need format so it isnt going to be more efficient or anything

Reply:
Clippy suggests replacing a good old for loop with a `fold()` call? I find that hard to believe.

Can you not keep the simplest possible procedural code?



> Also here, I think a `let mut result = String::new();` followed by a `for` loop is easier to understand.